diff --git a/luna/gateware/usb/usb2/device.py b/luna/gateware/usb/usb2/device.py
index 67e4c6b..4956cd3 100644
--- a/luna/gateware/usb/usb2/device.py
+++ b/luna/gateware/usb/usb2/device.py
@@ -232,7 +232,7 @@ class USBDevice(Elaboratable):
         m.d.comb += [
             reset_sequencer.bus_busy        .eq(self.bus_busy),
 
-            reset_sequencer.vbus_connected  .eq(~self.utmi.session_end),
+            reset_sequencer.vbus_connected  .eq(~self.utmi.session_end & self.connect),
             reset_sequencer.line_state      .eq(self.utmi.line_state),
         ]
 
